File: Commons/Action/src/main/java/com/greentree/commons/action/ContainerAction.java
package com.greentree.commons.action;

import com.greentree.commons.action.container.ListenerContainer;

/** @author Arseny Latyshev */
public abstract class ContainerAction<L, LC extends ListenerContainer<L>> implements AutoCloseable {

    protected final LC listeners;

    public ContainerAction(LC container) {
        this.listeners = container;
    }

    @Override
    public String toString() {
        String builder = "MultiAction [" +
                         listeners +
                         "]";
        return builder;
    }

    public ListenerCloser addListener(L listener) {
        return listeners.add(listener);
    }

    public int listenerSize() {
        return listeners.size();
    }

    @Override
    public void close() throws Exception {
        clear();
    }

    public void clear() {
        listeners.clear();
    }

}

File: Commons/Event/src/main/java/com/greentree/commons/event/ContainerAction.kt
package com.greentree.commons.event

import com.greentree.commons.event.container.ListenerContainer
import com.greentree.commons.event.container.MultiContainer

open class ContainerAction<L>(
	val container: ListenerContainer<L> = MultiContainer()
) : AutoCloseable {

	fun addListener(listener: L): ListenerCloser {
		return container.add(listener)
	}

	override fun close() {
		container.clear()
	}
}
